README.md
<p align="center">
    <img src="./assets/icon.png">
</p>

# epick

[![Build Status](https://github.com/vv9k/epick/workflows/epick%20CI/badge.svg)](https://github.com/vv9k/epick/actions?query=workflow%3A%22epick+CI%22)

Color picker for creating harmonic color palettes that works on Linux, Windows, macOS and web.

## Get it

You can checkout the web demo over [here](https://vv9k.github.io/epick/) or get a native binary from the [GitHub release page](https://github.com/vv9k/epick/releases).

If you happen to use Arch Linux you can grab **epick** from [AUR](https://aur.archlinux.org/packages/epick/) using your favorite package manager:
```bash
$ paru -S epick
```

## Build

Install required libraries (only required on Linux):
```
$ apt-get install libxcb-render0-dev libxcb-shape0-dev libxcb-xfixes0-dev libxkbcommon-dev
```

To manually build **epick** you'll need the latest Rust with `cargo`. To build the project run:
```
$ make
```

To start the web version locally run:
```
$ make start_web
```
This will build the WASM files and start a simple http server listening at `127.0.0.1:8080`.

To build without `make` checkout the build instructions in the `Makefile`

## Demo

To checkout the latest build of master branch head over to the [web demo](https://vv9k.github.io/epick).

## Keyboard shortcuts

Here are some handy shortcuts to enhance the usage of **epick**:
- Only supported with screen picker:
   - `p` to pick a color from under the cursor
   - `s` to save a color from under the cursor
- Other:
   - `h` toggle side panel

## Custom color format

To fully customize the way the colors are presented textually on the screen and the way they are copied to clipboard, a
simple formatting language has been introduced that resembles the syntax of rust formatting macros. Here is a simple
example of such a formatting string:
```
"{r} {g} {b}"
// Same as
"{ r } {  g} {    b     }"
```

Above example shows how to display a value of a field. The name of the field is enclosed in curly braces and can contain
multiple white space characters following the opening brace and preceeding the closing brace. This format string will
print red, green and blue values of the color in the 0.0 ..= 1.0 range.

To specify the precision of a floating point number:
```
"{r:.2} {g:.0} {b:.4}"
```

To display a value in decimal, octal, hex or uppercase hex respectively:
```
"{lab_l:d} {r255:o} {g255:x} {b255:X}"
```
Using this flags on floating values will automatically truncate the fractional part and treat the number as an integer.

### Supported color fields:
| Field       | Color value    | Value range      |
|-------------|----------------|------------------|
| `r`         | Red            | 0.0 ..= 1.0      |
| `g`         | Green          | 0.0 ..= 1.0      |
| `b`         | Blue           | 0.0 ..= 1.0      |
| `r255`      | Red            | 0 ..= 255        |
| `g255`      | Green          | 0 ..= 255        |
| `b255`      | Blue           | 0 ..= 255        |
| `cmyk_c`    | Cyan           | 0.0 ..= 1.0      |
| `cmyk_m`    | Magenta        | 0.0 ..= 1.0      |
| `cmyk_y`    | Yellow         | 0.0 ..= 1.0      |
| `cmyk_k`    | Key            | 0.0 ..= 1.0      |
| `cmyk_c100` | Cyan           | 0.0 ..= 100.0    |
| `cmyk_m100` | Magenta        | 0.0 ..= 100.0    |
| `cmyk_y100` | Yellow         | 0.0 ..= 100.0    |
| `cmyk_k100` | Key            | 0.0 ..= 100.0    |
| `hsl_h`     | HSL Hue        | 0.0 ..= 1.0      |
| `hsl_s`     | HSL Saturation | 0.0 ..= 1.0      |
| `hsl_l`     | HSL Light      | 0.0 ..= 1.0      |
| `hsl_h360`  | HSL Hue        | 0.0 ..= 360.0    |
| `hsl_s100`  | HSL Saturation | 0.0 ..= 100.0    |
| `hsl_l100`  | HSL Light      | 0.0 ..= 100.0    |
| `hsv_h`     | HSV Hue        | 0.0 ..= 1.0      |
| `hsv_s`     | HSV Saturation | 0.0 ..= 1.0      |
| `hsv_v`     | HSV Value      | 0.0 ..= 1.0      |
| `hsv_h360`  | HSV Hue        | 0.0 ..= 360.0    |
| `hsv_s100`  | HSV Saturation | 0.0 ..= 100.0    |
| `hsv_v100`  | HSV Value      | 0.0 ..= 100.0    |
| `lab_l`     | Lab Light      | 0.0 ..= 100.0    |
| `lab_a`     | Lab a          | -127.0 ..= 128.0 |
| `lab_b`     | Lab b          | -127.0 ..= 128.0 |
| `luv_l`     | Luv Light      | 0.0 ..= 100.0    |
| `luv_u`     | Luv u          | -134.0 ..= 220.0 |
| `luv_v`     | Luv v          | -140.0 ..= 122.0 |
| `lch_ab_l`  | LCH(ab) Light  | 0.0 ..= 100.0    |
| `lch_ab_c`  | LCH(ab) Chroma | 0.0 ..= 270.0    |
| `lch_ab_h`  | LCH(ab) Hue    | 0.0 ..= 360.0    |
| `lch_uv_l`  | LCH(uv) Light  | 0.0 ..= 100.0    |
| `lch_uv_c`  | LCH(uv) Chroma | 0.0 ..= 270.0    |
| `lch_uv_h`  | LCH(uv) Hue    | 0.0 ..= 360.0    |
| `xyy_x`     | xyY x          |                  |
| `xyy_y`     | xyY y          |                  |
| `xyy_Y`     | xyY Y          |                  |
| `xyz_x`     | XYZ X          |                  |
| `xyz_y`     | XYZ Y          |                  |
| `xyz_z`     | XYZ Z          |                  |

## License
[GPLv3](https://github.com/vv9k/epick/blob/master/LICENSE)


CHANGELOG.md
#
- Add a way to scale the UI in the settings
- Display RGB slider values in 0-255 range as integers rather than floats
- Move harmonies box outside of the harmonies header
- Add a way to hide the harmonies color box
- Increase the default size of current color and color under cursor box

# 0.9.0
- Change button layout in palette view
- Bring back ability to select color by primary clicking it
- Add a way to define custom palette formats and use them to copy palettes to clipboard.
- Use rust fmt for formatting numbers (fixes formatting of floats < 0.1)

# 0.8.2
- Fixed wasm build

# 0.8.1
- Update dependencies to mitigate security issues

# 0.8.0
- Add an extra black border to the zoom picker window on Linux so that the window is visible on both black and white backgrounds
- Remove sleep when zoom window is dragged so that the movement in smoother
- Fix side panel width when no colors are added
- Add support for multiple palettes
- Disable keybindings when entering text
- Add new central panel tab where all palettes are displayed
- Remove borders from color boxes
- Fix scrollbar position in picker ui

# 0.7.0
- Fix invisible circle on 2D HSV slider
- Update layout of current and cursor color [#40](https://github.com/vv9k/epick/pull/40)
- Zoom window can now be enabled by dragging the zoom button, this way the main window stays in focus and colors from other windows can be extracted. [#41](https://github.com/vv9k/epick/pull/41)
- Set app icon on OS taskbar
- Revert back to default display format when currently used custom format is deleted
- Fix program freeze when dragging a color [#45](https://github.com/vv9k/epick/pull/45)

# 0.6.0
- Add support for persistent configuration. A configuration file with a YAML syntax will be loaded on startup from appropriate config directory depending on OS.
  Supported options are the same as in settings window. The settings window now contains a `save settings` button.
- Persist colors throughout launches. On exit all current colors will be saved to a cache file and loaded back on startup.
- Add option to disable color caching
- Add keyboard shortcuts to display zoomed window, pick and save the color under the cursor
- Fix color rendering
- Add display picker on macOS
- Add shortcut to toggle side panel
- Add a help window with keybindings
- Fix circle of zoomed display picker on X11, the circle will now correctly point to a single pixel rather than four at a time.
- Add setting to set default color harmony
- Add 2D slider to HSV
- Add persistent colors and settings to WASM by using local browser storage
- All windows will now open in the empty area after sliders so that nothing is hidden
- Use different cursor when hovering over different elements
- Add custom color format support, you can now fully customise the way the colors next to color boxes are displayed.
  A simple formatting language has been introduced that has a syntax simillar to Rust's format macros.
- Detect dark mode OS preference when available and set the correct visuals
- Store theme preference in persistent configuration
- Add ability to save custom color display formats
- Add a checkbox to disable color labels in harmonies view
- Fix double `#` when copying a list of hex colors
- Add different harmony layouts like `square`, `line` or `stacked`
- Add gradient view of harmony
- Hide all sliders when opening the app

# 0.5.1
- Fix switching between working spaces and illuminants

# 0.5.0
- Add ability to spawn a window that will follow the cursor and display a zoomed in image with a pointer to a pixel (Only X11 and Windows)
- Make RGB and CMYK sliders adjust color when values change
- Add different display formats like: 'css rgb', 'css hsl' and previously available 'hex', 'hex uppercase'
- Fix hex input conversion to RGB
- Add settings to disable each colorspace [#13](https://github.com/vv9k/epick/pull/13)
- Updated light theme colors
- Add square colors
- Add monochromatic colors
- Rename `schemes` to `harmonies`
- CMYK sliders are now scaled from 0 ..= 100 [#14](https://github.com/vv9k/epick/pull/14)
- HSV sliders are now scaled 0..=360, 0..=100, 0..=100 respectively [#14](https://github.com/vv9k/epick/pull/14)
- HSL sliders are now scaled 0..=360, 0..=100, 0..=100 respectively [#14](https://github.com/vv9k/epick/pull/14)
- The triangle displaying current position on a slider will be displayed in the correct position [#14](https://github.com/vv9k/epick/pull/14)
- Add CIE Luv and CIE LCH(uv) sliders
- Add setting to select RGB working space
- Add CIE Lab and CIE LCH(ab) sliders
- The sidebar and topbar icons are now on the right side
- Fix invisible pointer on slider when using light theme
- Shades, tints and hues can now all be open at the same time
- Add settings to change the reference white illuminant
- The harmonies section is now on the top
- Fix sliders behaviour in the 0.0 ..= 1.0 range

# 0.4.0

- When moving the value slider of HSV or key slider of CMYK to edge values - 0 and 1 respectively, all other values would get reset to 0. Now the values are saved and when the value or key is brought back a little bit the values are restored
- Add HSL slider
- Windows opened by tabs like `hues`, `tints` or `shades` are now free to move around
- Monospaced text now correctly uses a custom font `FiraCode` for rendering
- Support for displaying color under the cursor (X11 and Windows supported for now)
- Added Windows and MacOS builds to GitHub releases
- Use a native-dialog when selecting the path to export the saved colors palette


src/app.rs
use crate::color::{color_as_hex, parse_color, Cmyk, Gradient};
use egui::color::*;
use egui::{
    pos2, vec2, Button, ImageButton, Rect, Response, ScrollArea, Slider, TextStyle, TextureId, Ui,
    Vec2,
};
use std::collections::HashMap;

use clipboard::ClipboardContext;
use clipboard::ClipboardProvider;

#[derive(Default)]
pub struct Epick {
    pub tab: EpickApp,
    pub picker: ColorPicker,
    pub saved_colors: Vec<(String, Color32)>,
}

pub enum EpickApp {
    ColorPicker,
    GradientView,
}

impl Default for EpickApp {
    fn default() -> Self {
        Self::ColorPicker
    }
}

pub struct ColorPicker {
    pub hex_color: String,
    pub cur_color: Option<Color32>,
    pub cur_hsva: Option<Hsva>,
    pub cur_cmyk: Option<Cmyk>,
    pub r: u8,
    pub g: u8,
    pub b: u8,
    pub hue: f32,
    pub sat: f32,
    pub val: f32,
    pub c: f32,
    pub m: f32,
    pub y: f32,
    pub k: f32,
    pub tex_mngr: TextureManager,
}

impl Default for ColorPicker {
    fn default() -> Self {
        Self {
            hex_color: "".to_string(),
            cur_color: Some(Color32::BLACK),
            cur_hsva: Some(Hsva::new(0., 0., 0., 1.)),
            cur_cmyk: Some(Cmyk::from(Color32::BLACK)),
            r: 0,
            g: 0,
            b: 0,
            hue: 0.,
            sat: 0.,
            val: 0.,
            c: 0.,
            m: 0.,
            y: 0.,
            k: 0.,
            tex_mngr: TextureManager::default(),
        }
    }
}

fn save_to_clipboard(text: String) -> Result<(), Box<dyn std::error::Error>> {
    let mut ctx: ClipboardContext = ClipboardProvider::new()?;
    ctx.set_contents(text)
}

impl epi::App for Epick {
    fn name(&self) -> &str {
        "epick"
    }

    fn update(&mut self, ctx: &egui::CtxRef, frame: &mut epi::Frame<'_>) {
        let _frame = egui::Frame {
            fill: Color32::from_rgb(22, 28, 35),
            margin: vec2(20., 20.),
            ..Default::default()
        };

        let _dark_frame_side = egui::Frame {
            fill: Color32::from_rgb(17, 22, 27),
            margin: vec2(15., 10.),
            ..Default::default()
        };

        let _dark_frame_top = egui::Frame {
            fill: Color32::from_rgb(17, 22, 27),
            margin: vec2(5., 5.),
            ..Default::default()
        };

        egui::TopPanel::top("top panel")
            .frame(_dark_frame_top.clone())
            .show(ctx, |ui| {
                self.top_ui(ui);
            });

        egui::SidePanel::left("colors", 150.)
            .frame(_dark_frame_side)
            .show(ctx, |ui| {
                ScrollArea::auto_sized().show(ui, |ui| {
                    self.side_ui(ui, &mut Some(frame.tex_allocator()));
                })
            });

        egui::CentralPanel::default()
            .frame(_frame)
            .show(ctx, |ui| match self.tab {
                EpickApp::ColorPicker => {
                    self.picker
                        .ui(ui, &mut Some(frame.tex_allocator()), &mut self.saved_colors);
                }
                EpickApp::GradientView => {}
            });

        frame.set_window_size(ctx.used_size());
    }
}

enum EpickButton {
    Active,
    Inactive,
}

impl EpickButton {
    fn text_color() -> Color32 {
        Color32::from_rgb(229, 222, 214)
    }

    fn bg_color(&self) -> Color32 {
        match self {
            EpickButton::Active => Color32::from_rgb(49, 63, 78),
            EpickButton::Inactive => Color32::from_rgb(35, 45, 56),
        }
    }

    fn btn<T: AsRef<str>>(&self, text: T) -> Button {
        Button::new(text.as_ref())
            .text_color(Self::text_color())
            .fill(Some(self.bg_color()))
    }
}

impl Epick {
    pub fn top_ui(&mut self, ui: &mut Ui) {
        ui.horizontal(|ui| {
            self.dark_light_switch(ui);
            ui.label("switch ui color");
            ui.add_space(50.);
            let picker_tab;
            const PICKER_TITLE: &str = "picker";
            match self.tab {
                EpickApp::ColorPicker => {
                    picker_tab = EpickButton::Active.btn(PICKER_TITLE);
                }
                EpickApp::GradientView => {
                    picker_tab = EpickButton::Inactive.btn(PICKER_TITLE);
                }
            }
            let picker_resp = ui.add(picker_tab);
            if picker_resp.clicked() {
                self.tab = EpickApp::ColorPicker;
            }

            let gradient_tab;
            const GRADIENT_TITLE: &str = "gradient";
            match self.tab {
                EpickApp::GradientView => {
                    gradient_tab = EpickButton::Active.btn(GRADIENT_TITLE);
                }
                EpickApp::ColorPicker => {
                    gradient_tab = EpickButton::Inactive.btn(GRADIENT_TITLE);
                }
            }
            let gradient_resp = ui.add(gradient_tab);
            if gradient_resp.clicked() {
                self.tab = EpickApp::GradientView;
            }
        });
    }

    pub fn dark_light_switch(&mut self, ui: &mut Ui) {
        let style = (*ui.ctx().style()).clone();
        let new_visuals = style.visuals.light_dark_small_toggle_button(ui);
        if let Some(visuals) = new_visuals {
            ui.ctx().set_visuals(visuals);
        }
    }

    pub fn side_ui(
        &mut self,
        ui: &mut Ui,
        tex_allocator: &mut Option<&mut dyn epi::TextureAllocator>,
    ) {
        ui.vertical(|ui| {
            ui.horizontal(|ui| {
                ui.heading("Saved colors");
                ui.add_space(7.);
                let clear_btn = ui.add(EpickButton::Inactive.btn("clear"));
                if clear_btn.clicked() {
                    self.saved_colors.clear();
                }
            });
            ui.add_space(7.);
            ui.label("Left click: set current");
            ui.add_space(3.5);
            ui.label("Right click: copy hex");
            ui.add_space(7.);

            for (idx, (hex, color)) in self.saved_colors.clone().iter().enumerate() {
                ui.horizontal(|ui| {
                    ui.vertical(|ui| {
                        ui.add_space(ui.fonts().row_height(TextStyle::Monospace));
                        ui.monospace(format!("#{}", hex));
                        ui.horizontal(|ui| {
                            if ui.button("❌").clicked() {
                                self.saved_colors
                                    .iter()
                                    .position(|(_hex, _)| _hex == hex)
                                    .map(|i| self.saved_colors.remove(i));
                            }
                            ui.vertical(|ui| {
                                if ui.button("⏶").clicked() {
                                    if idx > 0 {
                                        self.saved_colors.swap(idx, idx - 1);
                                    }
                                }

                                if ui.button("⏷").clicked() {
                                    if idx < (self.saved_colors.len() - 1) {
                                        self.saved_colors.swap(idx, idx + 1);
                                    }
                                }
                            });
                        });
                    });

                    let resp = tex_color(
                        ui,
                        tex_allocator,
                        &mut self.picker.tex_mngr,
                        color.clone(),
                        vec2(100., 50.),
                        Some(&hex),
                    );

                    if let Some(resp) = resp {
                        match self.tab {
                            EpickApp::ColorPicker => {
                                let hex = color_as_hex(&color);
                                if resp.clicked() {
                                    self.picker.set_cur_color(color.clone());
                                }

                                if resp.secondary_clicked() {
                                    let _ = save_to_clipboard(format!("#{}", hex));
                                }
                            }
                            EpickApp::GradientView => {}
                        };
                    }
                });
            }
        });
    }
}

impl ColorPicker {
    fn set_cur_color(&mut self, color: Color32) {
        self.r = color.r();
        self.g = color.g();
        self.b = color.b();
        let hsva = Hsva::from_srgb([self.r, self.g, self.b]);
        self.hue = hsva.h;
        self.sat = hsva.s;
        self.val = hsva.v;
        let cmyk = Cmyk::from(color);
        self.c = cmyk.c;
        self.m = cmyk.m;
        self.y = cmyk.y;
        self.k = cmyk.k;
        self.cur_color = Some(color);
        self.cur_hsva = Some(hsva);
        self.cur_cmyk = Some(cmyk);
    }

    pub fn ui(
        &mut self,
        ui: &mut Ui,
        tex_allocator: &mut Option<&mut dyn epi::TextureAllocator>,
        saved_colors: &mut Vec<(String, Color32)>,
    ) {
        ui.horizontal(|ui| {
            ui.label("Enter a hex color: ");
            let resp = ui.text_edit_singleline(&mut self.hex_color);
            if (resp.lost_focus() && ui.input().key_pressed(egui::Key::Enter))
                || ui.button("▶ go").clicked()
            {
                if let Some(color) = parse_color(self.hex_color.trim_start_matches("#")) {
                    self.set_cur_color(color);
                }
            }
            if ui.button("➕ save").clicked() {
                if let Some(color) = self.cur_color {
                    let color = (color_as_hex(&color), color);
                    if !saved_colors.contains(&color) {
                        saved_colors.push(color);
                    }
                }
            }
        });

        ui.add_space(15.);

        if let Some(color) = self.cur_color {
            ui.horizontal(|ui| {
                ui.label("Current color: ");
                ui.monospace(format!("#{}", color_as_hex(&color).to_uppercase()));
            });
            self.sliders(ui);
            ui.add_space(15.);

            if self.r != color.r() || self.g != color.g() || self.b != color.b() {
                self.set_cur_color(Color32::from_rgb(self.r, self.g, self.b));
            }

            // its ok to unwrap, cur_hsva is always set when cur_color is set
            let hsva = self.cur_hsva.unwrap();
            if self.hue != hsva.h || self.sat != hsva.s || self.val != hsva.v {
                let new_hsva = Hsva::new(self.hue, self.sat, self.val, 0.);
                let srgb = new_hsva.to_srgb();
                self.set_cur_color(Color32::from_rgb(srgb[0], srgb[1], srgb[2]));
            }

            let cmyk = self.cur_cmyk.clone().unwrap();
            if self.c != cmyk.c || self.m != cmyk.m || self.y != cmyk.y || self.k != cmyk.k {
                let new_cmyk = Cmyk::new(self.c, self.m, self.y, self.k);
                self.set_cur_color(Color32::from(new_cmyk));
            }

            ui.scope(|ui| {
                let resp = tex_color(
                    ui,
                    tex_allocator,
                    &mut self.tex_mngr,
                    color,
                    vec2(500., 500.),
                    Some(&color_as_hex(&color)),
                );
                if let Some(resp) = resp {
                    let hex = color_as_hex(&color);
                    if resp.clicked() {
                        self.set_cur_color(color);
                    }

                    if resp.secondary_clicked() {
                        let _ = save_to_clipboard(format!("#{}", hex));
                    }
                }
            });
        }
    }

    fn sliders(&mut self, ui: &mut Ui) {
        ui.horizontal(|ui| {
            ui.vertical(|ui| {
                ui.add_space(7.);
                ui.add(Slider::new(&mut self.r, u8::MIN..=u8::MAX).text("red"));
                ui.add_space(7.);
                ui.add(Slider::new(&mut self.g, u8::MIN..=u8::MAX).text("green"));
                ui.add_space(7.);
                ui.add(Slider::new(&mut self.b, u8::MIN..=u8::MAX).text("blue"));
            });
            ui.vertical(|ui| {
                ui.add_space(7.);
                ui.add(Slider::new(&mut self.hue, 0. ..=1.).text("hue"));
                ui.add_space(7.);
                ui.add(Slider::new(&mut self.sat, 0. ..=1.).text("saturation"));
                ui.add_space(7.);
                ui.add(Slider::new(&mut self.val, 0. ..=1.).text("value"));
            });
            ui.vertical(|ui| {
                ui.add_space(7.);
                ui.add(Slider::new(&mut self.c, 0. ..=1.).text("cyan"));
                ui.add_space(7.);
                ui.add(Slider::new(&mut self.m, 0. ..=1.).text("magenta"));
                ui.add_space(7.);
                ui.add(Slider::new(&mut self.y, 0. ..=1.).text("yellow"));
                ui.add_space(7.);
                ui.add(Slider::new(&mut self.k, 0. ..=1.).text("key"));
            });
        });
    }
}
fn tex_color(
    ui: &mut Ui,
    tex_allocator: &mut Option<&mut dyn epi::TextureAllocator>,
    tex_mngr: &mut TextureManager,
    color: Color32,
    size: Vec2,
    on_hover: Option<&str>,
) -> Option<Response> {
    let gradient = Gradient::one_color(color);
    let resp = tex_gradient(ui, tex_allocator, tex_mngr, &gradient, size, on_hover);

    resp
}

//self.set_cur_color(parse_color(&color).unwrap());
fn tex_gradient(
    ui: &mut Ui,
    tex_allocator: &mut Option<&mut dyn epi::TextureAllocator>,
    tex_mngr: &mut TextureManager,
    gradient: &Gradient,
    size: Vec2,
    on_hover: Option<&str>,
) -> Option<Response> {
    if let Some(tex_allocator) = tex_allocator {
        let resp = ui.horizontal(|ui| {
            let tex = tex_mngr.get(*tex_allocator, &gradient);
            let texel_offset = 0.5 / (gradient.0.len() as f32);
            let uv = Rect::from_min_max(pos2(texel_offset, 0.0), pos2(1.0 - texel_offset, 1.0));
            let image = ImageButton::new(tex, size).uv(uv);
            let mut resp = ui.add(image);

            if let Some(on_hover) = on_hover {
                resp = resp.on_hover_text(on_hover);
            }

            resp
        });
        return Some(resp.inner);
    }
    None
}

#[derive(Default)]
pub struct TextureManager(HashMap<Gradient, TextureId>);

impl TextureManager {
    fn get(
        &mut self,
        tex_allocator: &mut dyn epi::TextureAllocator,
        gradient: &Gradient,
    ) -> TextureId {
        *self.0.entry(gradient.clone()).or_insert_with(|| {
            let pixels = gradient.to_pixel_row();
            let width = pixels.len();
            let height = 1;
            tex_allocator.alloc_srgba_premultiplied((width, height), &pixels)
        })
    }
}
